File: FFI/linked_list/linked_list.rb
# C linked list in Ruby w/FFI

require 'ffi'

module LinkedList
	extend FFI::Library
	ffi_lib './liblinkedlist.so'

	class Node < FFI::Struct
		layout :data, :int,
			   :next, :pointer
	end


  # Attach the functions from the shared library
  attach_function :create_node, [:int], :pointer
  attach_function :append, [:pointer, :int], :void
  attach_function :get_data, [:pointer], :int
  attach_function :print_list, [:pointer], :void
  attach_function :free_list, [:pointer], :void
end

# Create the linked list and manipulate it
head = FFI::MemoryPointer.new(:pointer)

# Append some values
LinkedList.append(head, 10)
LinkedList.append(head, 20)
LinkedList.append(head, 30)

# Print the list
LinkedList.print_list(head.get_pointer(0))

# Get data from the first node
first_node = head.get_pointer(0)
puts "Data in the first node: #{LinkedList.get_data(first_node)}"

# Free the list
LinkedList.free_list(head.get_pointer(0))

File: FFI/linked_list/linked_list.c
// linked lists in C for Ruby via FFI

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
	int data;
	struct Node *next;
} Node;

Node* create_node(int data){
	Node * new_node = (Node*)malloc(sizeof(Node));

	if(new_node == NULL){
		perror("Malloc failed");
		exit(1);
	}

	new_node->data = data;
	new_node->next = NULL;
	return new_node;
}

void append(Node **head, int data){
	Node * new_node = create_node(data);
	if(*head == NULL){
		*head = new_node; 
	} else {

	Node * temp = *head;
	while(temp->next != NULL){
		temp = temp->next;
	}
	// we have reached the end of the list
	temp->next = new_node;
	}
}

int get_data(Node *node){
	return node->data;
}

void print_list(Node *head){
	Node *temp = head;

	while(temp != NULL){
		printf("%d -> ", temp->data);
		temp = temp->next;
	}
	printf("NULL\n");
}

void free_list(Node *head){
	Node *current = head;
	Node *next;

	while(current != NULL){
		next = current->next;
		free(current);
		current = next;
	}
}